### ISSUE TITLE
docs needed for nn_device

### ISSUE THREAD

**Issue by reqshark** [nickdesaulniers/node-nanomsg#145]
The nanomsg device api is useful should be documented

Related to #143, we have a device implementation and some tests for it, but none of it is documented.

**Comment by yoshuawuyts** [comment]
I took a lil stab at this today, but found I'm still lacking the complete mental models to describe it aptly. Perhaps in a few months I'll be able to revisit this, but I think it'd be preferred if someone else could document this.

**Comment by reqshark** [comment]
@yoshuawuyts see #147, you know we ought to get this documented

**Comment by yoshuawuyts** [comment]
yeah, definitely. But I haven't gone deep enough on this to do it quite yet. I'm going to be give a talk about `node-nanomsg` in a month, so I reckon by that time I will be able to do it - but if that's too late then someone will have to do it

**Comment by reqshark** [comment]
oh nice! We need more talks on `node-nanomsg`! :whale2: 

Also can you update this thread with any links about your talk? Or maybe we put a link to resources in the docs (oh but now that's a different PR)

Anyway yea a month is fine, if you see any issues with nn_device implementation also let us know so we can fix, I haven't tried using it personally in a long time

**Comment by yoshuawuyts** [comment]
hah, alright! I'll try and not forget to contribute back, don't hesitate to ping me if I haven't gravitated back here in a month. Cheers!

**Comment by reqshark** [comment]
nice talks notes, great diagrams.. like:
```txt
 ┌───────────┐
 │ Publisher │
 ├─────┬─────┤
 │ REP │ PUB │
 └▲──┬─┴──┬──┘
  1  2    3
 ┌┴──▼─┬──▼──┐
 │ REQ │ SUB │
 ├─────┴─────┤
 │Subscriber │
 └───────────┘
```
can we add stuff like that on the node-nanomsg readme? also my 3 cents: https://github.com/yoshuawuyts/talks/pull/2

**Comment by yoshuawuyts** [comment]
Oh yeah, that'd be cool! I've got a whole collection of 'em on https://github.com/yoshuawuyts/knowledge/blob/master/arch/patterns.md
